File: Content/Script/KDD/Core/KDDWidget.lua
-- ============================================================
-- KDDWidget.lua - Widget 的 Lua 侧包装基类
--
-- 继承 KDDView 的所有能力，额外提供：
--   1. 动画功能：自动识别 ShowAnim/HideAnim 并绑定播放
--   2. 自定义动画：Open/CustomClose/CustomLoop
--   3. 动画开关：bEnableAnimation（默认 true）
-- ============================================================

local KDDView = require("KDD.Core.KDDView")

local KDDWidget = {}
KDDWidget.__index = KDDWidget

-- 继承 KDDView 的方法
for k, v in pairs(KDDView) do
    KDDWidget[k] = v
end

function KDDWidget.New(Widget, WidgetName)
    local self = KDDView.New(Widget, WidgetName)
    setmetatable(self, KDDWidget)

    self.WidgetName = WidgetName
    self.bEnableAnimation = true

    -- 自动识别的动画名（蓝图侧需命名一致）
    self.AutoShowAnimName = "ShowAnim"
    self.AutoHideAnimName = "HideAnim"

    -- 自定义动画名（可外部设置）
    self.CustomShowAnimName = ""
    self.CustomLoopAnimName = ""
    self.CustomHideAnimName = ""

    return self
end

-- ======== 动画播放 ========

--- 获取蓝图中的 UWidgetAnimation 对象
function KDDWidget:_FindAnimation(AnimName)
    if not self.Content or not AnimName or AnimName == "" then return nil end
    -- 尝试通过 UMG 内置函数查找
    local OK, Anim = pcall(function()
        return self.Content:GetAnimation(AnimName)
    end)
    if OK and Anim then
        return Anim
    end
    return nil
end

--- 播放动画
function KDDWidget:_PlayAnim(Anim, StartAtTime, NumLoops, PlayMode)
    if not self.Content or not Anim then return end
    if not self.bEnableAnimation then return end
    local OK, Err = pcall(function()
        self.Content:PlayAnimation(Anim, StartAtTime or 0, NumLoops or 1, PlayMode or 0)
    end)
    if not OK then
        print("[KDDWidget] Play animation failed: " .. tostring(Err))
    end
end

--- 播放 Show 动画（自动识别 AutoShowAnimName）
function KDDWidget:PlayShowAnimation()
    if not self.bEnableAnimation then return end
    local Anim = self:_FindAnimation(self.AutoShowAnimName)
    if Anim then
        self:_PlayAnim(Anim, 0, 1)
        return
    end
    -- 回退到自定义动画
    if self.CustomShowAnimName and self.CustomShowAnimName ~= "" then
        local CustomAnim = self:_FindAnimation(self.CustomShowAnimName)
        if CustomAnim then
            self:_PlayAnim(CustomAnim, 0, 1)
        end
    end
end

--- 播放 Hide 动画（自动识别 AutoHideAnimName）
function KDDWidget:PlayHideAnimation()
    if not self.bEnableAnimation then return end
    local Anim = self:_FindAnimation(self.AutoHideAnimName)
    if Anim then
        self:_PlayAnim(Anim, 0, 1)
        return
    end
    -- 回退到自定义动画
    if self.CustomHideAnimName and self.CustomHideAnimName ~= "" then
        local CustomAnim = self:_FindAnimation(self.CustomHideAnimName)
        if CustomAnim then
            self:_PlayAnim(CustomAnim, 0, 1)
        end
    end
end

--- 播放 Loop 动画
function KDDWidget:PlayLoopAnimation()
    if not self.bEnableAnimation then return end
    if not self.CustomLoopAnimName or self.CustomLoopAnimName == "" then return end
    local Anim = self:_FindAnimation(self.CustomLoopAnimName)
    if Anim then
        -- Loop 模式：NumLoops = 0 表示无限循环
        self:_PlayAnim(Anim, 0, 0)
    end
end

--- 停止循环动画
function KDDWidget:StopLoopAnimation()
    if not self.Content then return end
    if not self.CustomLoopAnimName or self.CustomLoopAnimName == "" then return end
    local Anim = self:_FindAnimation(self.CustomLoopAnimName)
    if Anim then
        pcall(function()
            self.Content:StopAnimation(Anim)
        end)
    end
end

-- ======== 重写生命周期（注入动画） ========

function KDDWidget:OnOpen()
    self:PlayShowAnimation()
end

function KDDWidget:OnClose()
    self:PlayHideAnimation()
end

-- ======== 重写 Show/Hide（集成动画） ========

function KDDWidget:Show()
    if not self.Content then return end
    if self._isOpen then return end

    if not self._isCreated then
        self:OnCreate()
    end

    self.Content:AddToViewport()
    self._isOpen = true
    self:OnOpen()
end

function KDDWidget:Hide()
    if not self.Content then return end
    if not self._isOpen then return end

    self:OnClose()
    self.Content:RemoveFromParent()
    self._isOpen = false
end

-- ======== 可重写生命周期回调 ========

function KDDWidget:OnInit() end
function KDDWidget:OnShow() end
function KDDWidget:OnHide() end

-- ======== 设置可见性 ========

function KDDWidget:SetVisibility(Visibility)
    if self.Content then
        self.Content:SetVisibility(Visibility)
    end
end

-- ======== 设置自身为父控件的子控件 ========

function KDDWidget:AddToParent(ParentWidget)
    if self.Content and ParentWidget then
        self.Content:SetParent(ParentWidget)
    end
end

return KDDWidget

File: Source/GameDemo/UMG/KDDWidget.cpp
#include "KDDWidget.h"
#include "Blueprint/WidgetBlueprintGeneratedClass.h"

UKDDWidget::UKDDWidget(const FObjectInitializer& ObjectInitializer)
    : Super(ObjectInitializer)
    , bEnableAnimation(true)
    , AutoShowAnim(nullptr)
    , AutoHideAnim(nullptr)
    , AutoLoopAnim(nullptr)
{
}

FString UKDDWidget::GetModuleName_Implementation() const
{
    // ModuleLocator 会使用 KDDBindingManager 的配表查找 Lua 路径
    return TEXT("");
}

// ======== 生命周期 ========

void UKDDWidget::NativeConstruct()
{
    CacheAutoAnimations();
    Super::NativeConstruct();
    OnCreate();
    OnOpen();
}

void UKDDWidget::NativeDestruct()
{
    OnClose();
    OnDestroy();
    Super::NativeDestruct();
}

void UKDDWidget::NativeTick(const FGeometry& MyGeometry, float InDeltaTime)
{
    Super::NativeTick(MyGeometry, InDeltaTime);
}

// ======== 详细生命周期 ========

void UKDDWidget::OnCreate()
{
    UE_LOG(LogTemp, Log, TEXT("[UKDDWidget] OnCreate"));
}

void UKDDWidget::OnOpen()
{
    UE_LOG(LogTemp, Log, TEXT("[UKDDWidget] OnOpen"));
    PlayShowAnimation();
}

void UKDDWidget::OnClose()
{
    PlayHideAnimation();
    UE_LOG(LogTemp, Log, TEXT("[UKDDWidget] OnClose"));
}

void UKDDWidget::OnRefresh()
{
    UE_LOG(LogTemp, Log, TEXT("[UKDDWidget] OnRefresh"));
}

void UKDDWidget::OnDestroy()
{
    UE_LOG(LogTemp, Log, TEXT("[UKDDWidget] OnDestroy"));
}

// ======== 动画缓存 ========

void UKDDWidget::CacheAutoAnimations()
{
    AutoShowAnim = FindAnimationByName(TEXT("ShowAnim"));
    AutoHideAnim = FindAnimationByName(TEXT("HideAnim"));
    AutoLoopAnim = FindAnimationByName(TEXT("LoopAnim"));

    if (AutoShowAnim)
    {
        UE_LOG(LogTemp, Log, TEXT("[UKDDWidget] Auto-detected ShowAnim: %s"), *AutoShowAnim->GetName());
    }
    if (AutoHideAnim)
    {
        UE_LOG(LogTemp, Log, TEXT("[UKDDWidget] Auto-detected HideAnim: %s"), *AutoHideAnim->GetName());
    }
    if (AutoLoopAnim)
    {
        UE_LOG(LogTemp, Log, TEXT("[UKDDWidget] Auto-detected LoopAnim: %s"), *AutoLoopAnim->GetName());
    }
}

// ======== GetShowAnimation / GetHideAnimation ========

UWidgetAnimation* UKDDWidget::GetShowAnimation_Implementation() const
{
    // 优先自动识别的 ShowAnim
    if (AutoShowAnim) return AutoShowAnim;
    // 回退到自定义动画名
    if (!CustomShowAnimName.IsEmpty())
    {
        return const_cast<UKDDWidget*>(this)->FindAnimationByName(CustomShowAnimName);
    }
    // 兜底：直接按默认名查找
    return const_cast<UKDDWidget*>(this)->FindAnimationByName(TEXT("ShowAnim"));
}

UWidgetAnimation* UKDDWidget::GetHideAnimation_Implementation() const
{
    // 优先自动识别的 HideAnim
    if (AutoHideAnim) return AutoHideAnim;
    // 回退到自定义动画名
    if (!CustomHideAnimName.IsEmpty())
    {
        return const_cast<UKDDWidget*>(this)->FindAnimationByName(CustomHideAnimName);
    }
    // 兜底：直接按默认名查找
    return const_cast<UKDDWidget*>(this)->FindAnimationByName(TEXT("HideAnim"));
}

// ======== 动画播放 ========

void UKDDWidget::PlayShowAnimation()
{
    if (!bEnableAnimation) return;

    UWidgetAnimation* Anim = GetShowAnimation();
    if (Anim && IsValid(this))
    {
        PlayAnimation(Anim);
        UE_LOG(LogTemp, Log, TEXT("[UKDDWidget] Playing Show animation: %s"), *Anim->GetName());
    }
}

void UKDDWidget::PlayHideAnimation()
{
    if (!bEnableAnimation) return;

    UWidgetAnimation* Anim = GetHideAnimation();
    if (Anim && IsValid(this))
    {
        PlayAnimation(Anim);
        UE_LOG(LogTemp, Log, TEXT("[UKDDWidget] Playing Hide animation: %s"), *Anim->GetName());
    }
}

UWidgetAnimation* UKDDWidget::GetLoopAnimation() const
{
    // 优先自动识别的 LoopAnim
    if (AutoLoopAnim) return AutoLoopAnim;
    // 回退到自定义动画名
    if (!CustomLoopAnimName.IsEmpty())
    {
        return const_cast<UKDDWidget*>(this)->FindAnimationByName(CustomLoopAnimName);
    }
    // 兜底：直接按默认名查找
    return const_cast<UKDDWidget*>(this)->FindAnimationByName(TEXT("LoopAnim"));
}

void UKDDWidget::PlayLoopAnimation()
{
    if (!bEnableAnimation) return;

    UWidgetAnimation* Anim = GetLoopAnimation();
    if (Anim && IsValid(this))
    {
        PlayAnimation(Anim, 0.0f, 0);  // 0 = 无限循环
        UE_LOG(LogTemp, Log, TEXT("[UKDDWidget] Playing Loop animation: %s"), *Anim->GetName());
    }
}

void UKDDWidget::StopLoopAnimation()
{
    UWidgetAnimation* Anim = GetLoopAnimation();
    if (Anim && IsValid(this))
    {
        StopAnimation(Anim);
        UE_LOG(LogTemp, Log, TEXT("[UKDDWidget] Stopped Loop animation: %s"), *Anim->GetName());
    }
}

// ======== 按名称查找动画 ========

UWidgetAnimation* UKDDWidget::FindAnimationByName(const FString& AnimName) const
{
    if (AnimName.IsEmpty()) return nullptr;

    // 1) 通过反射查找同名 FObjectProperty（蓝图生成的动画属性）
    FProperty* Prop = GetClass()->FindPropertyByName(*AnimName);
    if (Prop)
    {
        FObjectProperty* ObjProp = CastField<FObjectProperty>(Prop);
        if (ObjProp)
        {
            UWidgetAnimation* Anim = Cast<UWidgetAnimation>(ObjProp->GetObjectPropertyValue_InContainer(this));
            if (Anim) return Anim;
        }
    }

    // 2) 兜底：扫描 WidgetBlueprintGeneratedClass::Animations 按名字匹配
    if (UWidgetBlueprintGeneratedClass* BGClass = Cast<UWidgetBlueprintGeneratedClass>(GetClass()))
    {
        for (UWidgetAnimation* Anim : BGClass->Animations)
        {
            if (Anim && Anim->GetFName() == *AnimName)
            {
                return Anim;
            }
        }
    }

    return nullptr;
}

// ======== 编辑器测试按钮 ========
//
// ★ 原理（参考 RedApp KFUIWidget）：
//   UE5.8 UMG 蓝图编辑器中，CallInEditor 在"设计器预览 Widget 实例"上执行，
//   该实例有有效的 GetWorld()（Editor World）。在此实例上：
//   1. PlayAnimation(Anim) 正常创建 FWidgetAnimationState 加入 ActiveAnimations
//   2. 但设计时没有世界 tick，AnimationState 不会被驱动
//   3. 用 GetWorld()->TimerManager.SetTimer() 设一个0.017s定时器
//   4. 定时器回调中手动调用 EditorAnimState->Tick(dt) + Invalidate 强制重绘
//   5. 动画完成回调停止定时器

#if WITH_EDITOR
#include "Animation/WidgetAnimationState.h"

// ======== 世界定时器管理（RedApp KFUIWidget 方案） ========

void UKDDWidget::Editor_TickPreviewAnim()
{
    if (IsDesignTime() && GetWorld())
    {
        int64 NowTicks = FDateTime::Now().GetTicks();
        float dt = (float)(NowTicks - EditorPrevTicks) / ETimespan::TicksPerSecond;

        if (EditorAnimState.IsValid())
        {
            EditorAnimState->Tick(dt);
        }

        if (TSharedPtr<SWidget> CachedWidget = GetCachedWidget())
        {
            CachedWidget->Invalidate(EInvalidateWidget::Paint);
        }

        EditorPrevTicks = NowTicks;
    }
}

void UKDDWidget::Editor_StartPreviewAnimTimer()
{
    if (IsDesignTime() && GetWorld())
    {
        EditorPrevTicks = FDateTime::Now().GetTicks();
        GetWorld()->GetTimerManager().SetTimer(EditorPreviewTimerHandle, this, &UKDDWidget::Editor_TickPreviewAnim, 0.017f, true);
    }
}

void UKDDWidget::Editor_StopPreviewAnimTimer()
{
    if (IsDesignTime() && GetWorld())
    {
        GetWorld()->GetTimerManager().ClearTimer(EditorPreviewTimerHandle);
    }
}

void UKDDWidget::Editor_OnAnimFinished()
{
    EditorAnimState.Reset();
    Editor_StopPreviewAnimTimer();
}

// ======== CallInEditor 入口 ========

void UKDDWidget::Editor_PlayShowAnim()
{
    UE_LOG(LogTemp, Log, TEXT("[KDDWidget] ▶ Show: this=%s World=%s"), *GetName(),
        GetWorld() ? *GetWorld()->GetName() : TEXT("null"));

    UWidgetAnimation* Anim = GetShowAnimation();
    if (!Anim)
    {
        UE_LOG(LogTemp, Warning, TEXT("[KDDWidget] ShowAnim not found in blueprint"));
        return;
    }
    UE_LOG(LogTemp, Log, TEXT("[KDDWidget] ShowAnim found: %s"), *Anim->GetName());

    // 停止之前的预览
    Editor_StopPreviewAnimTimer();
    EditorAnimState.Reset();

    // 绑定完成回调
    UnbindFromAnimationFinished(Anim, EditorAnimFinishEvt);
    EditorAnimFinishEvt.BindDynamic(this, &UKDDWidget::Editor_OnAnimFinished);
    BindToAnimationFinished(Anim, EditorAnimFinishEvt);

    // 播放动画（创建动画状态）
    PlayAnimation(Anim, 0.0f, 1, EUMGSequencePlayMode::Forward, 1.0f, false);

    // 获取 FWidgetAnimationState 句柄用于手动 tick
    EditorAnimState = GetAnimationState(Anim);
    UE_LOG(LogTemp, Log, TEXT("[KDDWidget] AnimState valid=%d"), EditorAnimState.IsValid());

    // 启动世界定时器驱动动画
    Editor_StartPreviewAnimTimer();
    UE_LOG(LogTemp, Log, TEXT("[KDDWidget] Preview timer started"));
}

void UKDDWidget::Editor_PlayLoopAnim()
{
    UE_LOG(LogTemp, Log, TEXT("[KDDWidget] ▶ Loop: this=%s"), *GetName());

    UWidgetAnimation* Anim = GetLoopAnimation();
    if (!Anim)
    {
        UE_LOG(LogTemp, Warning, TEXT("[KDDWidget] LoopAnim not found"));
        return;
    }

    Editor_StopPreviewAnimTimer();
    EditorAnimState.Reset();

    // 循环动画不需要完成回调
    PlayAnimation(Anim, 0.0f, 0, EUMGSequencePlayMode::Forward, 1.0f, false);
    EditorAnimState = GetAnimationState(Anim);

    Editor_StartPreviewAnimTimer();
}

void UKDDWidget::Editor_PlayHideAnim()
{
    UE_LOG(LogTemp, Log, TEXT("[KDDWidget] ▶ Hide: this=%s"), *GetName());

    UWidgetAnimation* Anim = GetHideAnimation();
    if (!Anim)
    {
        UE_LOG(LogTemp, Warning, TEXT("[KDDWidget] HideAnim not found"));
        return;
    }
    UE_LOG(LogTemp, Log, TEXT("[KDDWidget] HideAnim found: %s"), *Anim->GetName());

    Editor_StopPreviewAnimTimer();
    EditorAnimState.Reset();

    UnbindFromAnimationFinished(Anim, EditorAnimFinishEvt);
    EditorAnimFinishEvt.BindDynamic(this, &UKDDWidget::Editor_OnAnimFinished);
    BindToAnimationFinished(Anim, EditorAnimFinishEvt);

    PlayAnimation(Anim, 0.0f, 1, EUMGSequencePlayMode::Forward, 1.0f, false);
    EditorAnimState = GetAnimationState(Anim);

    Editor_StartPreviewAnimTimer();
}
#endif
